<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Algorithme de Newton </h1>
</center>

## Implémentation 
 
1. Coder l’algorithme de Newton dans le fichier `src/newton.jl` en respectant la spécification donnée dans ce même fichier ;
2. Exécuter les tests ci-dessous et vérifier qu'ils passent.

Pour les tests, nous avons défini les fonctions suivantes $f_1 \colon \mathbb{R}^3 \to \mathbb{R}$
et $f_2 \colon \mathbb{R}^2 \to \mathbb{R}$.

$$
    f_{1}(x_1, x_2, x_3) = 2 (x_1 +x_2 + x_3 -3)^2 + (x_1-x_2)^2 + (x_2 - x_3)^2
$$
et
$$
    f_{2}(x_1,x_2) = 100(x_2-x_1^2)^2 + (1-x_1)^2.
$$

**Remarque.** On peut retrouver ces fonctions dans le fichier `test/fonctions_de_tests.jl`.

In [1]:
include("../src/newton.jl")         # votre algorithme de Newton
include("../test/tester_newton.jl") # la fonction pour tester votre algorithme de Newton

#
afficher = true # si true, alors affiche les résultats des algorithmes

#
tester_newton(newton, afficher); # tester l'algorithme de Newton

Affichage des résultats des algorithmes : true

-------------------------------------------------------------------------
Résultats de : newton appliqué à f1:
  * x0       = [1, 1, 1]
  * x_sol    = [1, 1, 1]
  * f(x_sol) = 0
  * nb_iters = 0
  * flag     = 0
  * solution = [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : newton appliqué à f1:
  * x0       = [1, 0, 0]
  * x_sol    = [1.0, 1.0, 0.9999999999999999]
  * f(x_sol) = 1.232595164407831e-32
  * nb_iters = 1
  * flag     = 0
  * solution = [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : newton appliqué à f1:
  * x0       = [10.0, 3.0, -2.2]
  * x_sol    = [1.0, 0.9999999999999996, 0.9999999999999987]
  * f(x_sol) = 7.296963373294359e-30
  * nb_iters = 1
  * flag     = 0
  * solution = [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : newton appliqué à f2:
  * x0       = [-1.2, 1.0

In [2]:
include("../src/newton.jl") # votre algorithme de Newton
include("../test/fonctions_de_tests.jl") # pour avoir la fonction d'affichage des résultats

# Fonction f0
# -----------
f0(x)      =  sin(x)
grad_f0(x) =  cos(x) # la gradient de la fonction f0
hess_f0(x) = -sin(x) # la hessienne de la fonction f0
solution   = -pi/2

x0 = solution
x_sol, f_sol, flag, nb_iters = newton(f0, grad_f0, hess_f0, x0)
afficher_resultats("Newton", "f0", x0, x_sol, f_sol, flag, nb_iters, solution)

x0 = -pi/2+0.5
x_sol, f_sol, flag, nb_iters = newton(f0, grad_f0, hess_f0, x0)
afficher_resultats("Newton", "f0", x0, x_sol, f_sol, flag, nb_iters, solution)

x0 = pi/2
x_sol, f_sol, flag, nb_iters = newton(f0, grad_f0, hess_f0, x0)
afficher_resultats("Newton", "f0", x0, x_sol, f_sol, flag, nb_iters, solution)


-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0:
  * x0       = -1.5707963267948966
  * x_sol    = -1.5707963267948966
  * f(x_sol) = -1.0
  * nb_iters = 0
  * flag     = 0
  * solution = -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0:
  * x0       = -1.0707963267948966
  * x_sol    = -1.5707963267949088
  * f(x_sol) = -1.0
  * nb_iters = 3
  * flag     = 0
  * solution = -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0:
  * x0       = 1.5707963267948966
  * x_sol    = 1.5707963267948966
  * f(x_sol) = 1.0
  * nb_iters = 0
  * flag     = 0
  * solution = -1.5707963267948966


## Interprétation 

1. Justifier les résultats obtenus pour l'exemple $f_0$ ci-dessus;
2. Justifier que l’algorithme implémenté converge en une itération pour $f_{1}$;
3. Justifier que l’algorithme puisse ne pas converger pour $f_{2}$ avec certains points initiaux.

**Remarque.** Vous pouvez mettre `affiche=true` dans les tests de l'algorithme de Newton pour
vous aider.


## Justifications

#### Q1.
.Dans le premier cas, aucune itération n'est nécessaire car le minimum est identique à la valeur initiale $x_{0}$.

.Dans le deuxième cas, la méthode converge après 3 itérations (pas beaucoup car le $x_{0}$ est proche de la solution exacte)  vers une valeur ($-1.5707963267949088$) =~ la solution exacte (la valeur n'est pas precisèment égale à la solution exacte à cause de la marge d'erreur donné par la tolérance absolue).

.Dans le dernier cas, $x_{0} = \frac{\pi}{2}$ annule le $gradf$ donc on ne rentre pas dans la boucle d'où la justification de $nbiters = 0$ , or $\frac{\pi}{2}$ est un maximum non pas un minimum de la fonction $sin$, donc on peut citer que si l'algorithme trouve un point critique qui ne represente pas un minimum c'est terminé.

#### Q2.
L'algorithme converge en une seule itération pour la fonction $f_{1}$ parce qu'elle est quadratique et donc identique à son développement de Taylor à l'ordre 2. En conséquence, la solution est précisément(avec tolérance près) déterminée après la première itération au max, étant donné que la méthode de Newton converge rapidement pour les fonctions quadratiques en se basant sur les caractéristiques du développement de Taylor.Taylor.





#### Q3.
Pour certains points initiaux , on ne peut pas calculer le $d_{k}$ car la hessienne en $x_{0}$ n'est pas inversible , dans ce cas $H(f_{2})$ = \begin{bmatrix}
    1200x_1^2 - 400x_2 + 2 & -400x_1 \\
    -400x_1 & 2 0
\end{bmatrix}

i
x}
|


pour $x_{0} = [0 - \frac{1}{200}] = [0 - 0.005]$(comme dans le test) on aura $det(H(f_{2})(x_{0})) = 0$ ce qui va arrêter le programme dans la première itération.